### DISCUSSION TITLE
RFC: Simplification of Pathname Navigation API in TanStack Router

### DISCUSSION THREAD

**Discussion by tannerlinsley** [TanStack/router#1277] [Ideas]
## Summary

This RFC proposes a simplification of the `pathname` navigation API within the TanStack Router to enhance developer experience and type safety. The proposed changes aim to streamline the API by eliminating the need for separate `to` and `params` in navigation, instead supporting direct interpolation within the `to` string.

## Motivation

The current API, while designed for type safety and ease of use through autocomplete suggestions and strict dynamic path parameters, has been identified as somewhat clunky and limiting. The complexity arises especially when dealing with dynamic path parameters, leading to a less intuitive API.

## Proposal

The new API version will require only the `to` field for navigation, supporting dynamic path interpolation directly. This approach will retain autocomplete functionality and type safety while making the API more straightforward to use. The main features of the proposed API include:

- **Simplified Navigation**: Developers can directly interpolate dynamic path values into the `to` string, reducing the API's complexity.
- **Autocomplete Templates**: Enhance the developer experience by providing autocomplete suggestions based on the route being typed.
- **Type Safety with Dynamic Paths**: When a path containing dynamic segments (e.g., `/posts/$postId/`) is detected, the system will enforce type safety by requiring the interpolation of actual values or variables.

### Example Usage

Upon typing `/post`, a developer would see suggestions like:

- `/posts`
- `/posts/$postId/`
- `/posts/$postId/edit`

Selecting `/posts/$postId/` would then prompt the developer to interpolate a value for `$postId`, ensuring type safety.

### Breaking Changes

- **Elimination of Parse/Serialize for Path Params**: This proposal suggests removing the parse/serialize functionality for path parameters, which is considered an unfavorable pattern. This change streamlines the API and focuses on direct value interpolation.

### Continued Support for `params` API

The `params` API will still be available as a method for updating the current route's parameters. This ensures backward compatibility and flexibility for cases where direct `to` interpolation may not be used.

## Rationale

This proposal stems from a need to simplify the TanStack Router's navigation API, making it more intuitive and less error-prone. By enabling direct interpolation within the `to` string and removing less-desirable patterns like parse/serialize for path parameters, we can offer a more streamlined and developer-friendly API.

## Implementation

The proposed changes will require updates to the TanStack Router's core navigation logic to support direct path interpolation and enforce type safety dynamically. Additionally, the autocomplete suggestion mechanism will be enhanced to provide more relevant templates based on the developer's input.

## Open Questions

- **Migration Path**: How can we ensure a smooth transition for existing users of the TanStack Router to this new API model?
- **Type Safety Enforcement**: What mechanisms will be employed to detect and enforce type safety dynamically based on interpolated path values?
- **Impact on Existing Features**: How will these changes affect other router features, such as route matching and link generation?

## Feedback

Community feedback is crucial to refining and validating this proposal. We welcome comments, concerns, and suggestions from all stakeholders to ensure that the new API meets the needs of the broader developer community.

---

This RFC is open for discussion and feedback. Please share your thoughts and suggestions to help shape the future of pathname navigation in the TanStack Router.

**Comment by chorobin** [comment]
I'm wondering about suggestions. I'm not sure how good suggestions for template literal types are now.

For example:

`/posts/$postId`

Needs to be replaced with:

`/posts/post1`

So I assume this is a type checker error. Which will work fine.

How does the ide help the dev reach this conclusion as I'm not sure if suggestions work for this.

A `params` object always shows the missing property you need to provide. Where I'm not sure if this is possible with template literal types as they might have worse suggestions.

**Comment by SeanCassiere** [comment]
Just two things on my side:

1. Whenever you open the user to having to build their strings, it feels like something that's just waiting to go wrong.
```tsx
const postId = '123'
<Link
  // oh no! my prettier broke stuff to a new line
  to={`/posts${
    postId    
  }/edit`}
  // would this bork my type-inference now?
/>
```

2. The `parseParams` and `stringifyParams` kept the door open for people wanting to run a basic base64 obfuscation on their ids that was handled by the router (which kind-of feels like the right place for something like this). This logic could ofc, be moved to the routeContext.
```
/posts/1  ->  /posts/YQ==
```

Objectively my only real concern is (**1.**), and would we be opening up the Router to weird string interpolation stuff that we don't really have any control over.

My second point was just a use-case I saw that Router made possible, but is not really any cause for concern.
